csr_check.go

```diff
@@ -162,3 +180,74 @@ func authorizeCSR(machineList *v1beta1.MachineList, req *certificatesv1beta1.Cer
 
 	return nil
 }
+
+func authorizeNodeClientCSR(machines []v1beta1.Machine, nodes corev1client.NodeInterface, req *certificatesv1beta1.CertificateSigningRequest, csr *x509.CertificateRequest) error {
+	if !isReqFromNodeBootstrapper(req) {
+		return fmt.Errorf("CSR %s for node client cert has wrong user", req.Name)
+	}
+
+	nodeName := strings.TrimPrefix(csr.Subject.CommonName, nodeUserPrefix)
+
+	_, err := nodes.Get(nodeName, metav1.GetOptions{})
+	switch {
+	case err == nil:
+		return fmt.Errorf("node %s already exists", nodeName)
+	case errors.IsNotFound(err):
+		// good, node does not exist
+	default:
+		return fmt.Errorf("failed to check if node %s already exists: %v", nodeName, err)
+	}
+
+	nodeMachine, ok := findMatchingMachineFromInternalDNS(nodeName, machines)
+	if !ok {
+		return fmt.Errorf("failed to find machine for node %s", nodeName)
+	}
+
+	if nodeMachine.Status.NodeRef != nil {
+		return fmt.Errorf("machine for node %s already has node ref", nodeName)
+	}
```

Comment:
Does finding the nodeMachine and checking for an empty ref satisfy: 5) Ensure there’s machine object that triggered this new node trying to join the cluster. ?

Reply:
I would say `findMatchingMachineFromInternalDNS` does that by itself.

Reply:
(granted you could say all of the machine checks are required to fulfill the requirement)